src/periphery/MagicGmRouter.sol

```diff
@@ -0,0 +1,264 @@
+    function containsBytes32(bytes32 setKey, bytes32 value) external view returns (bool);
+}
+
+struct MagicGmRouterOrderParams {
+    uint256 usdcAmount;
+    uint256 executionFee;
+    uint256 minMarketTokens;
+}
+
+contract MagicGmRouterOrder is IMagicGmRouterOrder {
+    using SafeTransferLib for address;
+
+    error ErrFinalized();
+    error ErrNotOnwer();
+    error ErrAlreadyInitialized();
+
+    bytes32 public constant DEPOSIT_LIST = keccak256(abi.encode("DEPOSIT_LIST"));
+
+    IMagicGm public immutable MAGIC_GM;
+    address public immutable USDC;
+    address public immutable GM_BTC;
+    address public immutable GM_ETH;
+    address public immutable GM_ARB;
+    address public immutable WBTC;
+    address public immutable WETH;
+    address public immutable ARB;
+    IGmxV2ExchangeRouter public immutable GMX_ROUTER;
+    IGmxDataStore public immutable DATASTORE;
+    address public immutable DEPOSIT_VAULT;
+    address public immutable SYNTHETICS_ROUTER;
+
+    address public owner;
+    bytes32 public btcOrderKey;
+    bytes32 public ethOrderKey;
+    bytes32 public arbOrderKey;
+
+    bool public finalized;
+
+    modifier onlyOwner() virtual {
+        if (msg.sender != owner) {
+            revert ErrNotOnwer();
+        }
+        _;
+    }
+
+    constructor(
+        IMagicGm _magicGm,
+        address _usdc,
+        address _gmBTC,
+        address _gmEth,
+        address _gmArb,
+        address _wbtc,
+        address _weth,
+        address _arb,
+        IGmxV2ExchangeRouter _gmxRouter,
+        address _syntheticsRouter
+    ) {
+        MAGIC_GM = _magicGm;
+        USDC = _usdc;
+        GM_BTC = _gmBTC;
+        GM_ETH = _gmEth;
+        GM_ARB = _gmArb;
+        WBTC = _wbtc;
+        WETH = _weth;
+        ARB = _arb;
+        GMX_ROUTER = _gmxRouter;
+        SYNTHETICS_ROUTER = _syntheticsRouter;
+        DATASTORE = IGmxDataStore(_gmxRouter.dataStore());
+        DEPOSIT_VAULT = IGmxV2DepositHandler(_gmxRouter.depositHandler()).depositVault();
+    }
+
+    function init(address _owner, MagicGmRouterOrderParams[] memory _params) external payable {
+        if (owner != address(0)) {
+            revert ErrAlreadyInitialized();
+        }
+
+        owner = _owner;
+
+        uint256 usdcBalance = USDC.balanceOf(address(this));
+        address(USDC).safeApprove(address(SYNTHETICS_ROUTER), usdcBalance);
+
+        btcOrderKey = _createDepositOrder(GM_BTC, WBTC, _params[0].usdcAmount, _params[0].minMarketTokens, _params[0].executionFee);
+        ethOrderKey = _createDepositOrder(GM_ETH, WETH, _params[1].usdcAmount, _params[1].minMarketTokens, _params[1].executionFee);
+        arbOrderKey = _createDepositOrder(GM_ARB, ARB, _params[2].usdcAmount, _params[2].minMarketTokens, _params[2].executionFee);
+    }
+
+    function claim() public onlyOwner {
+        if (finalized) {
+            revert ErrFinalized();
+        }
+
+        address(GM_ETH).safeApprove(address(MAGIC_GM), GM_ETH.balanceOf(address(this)));
```

Comment:
can we move this into constructor or init?

Reply:
yes but just need to max it but shouldn't matter much because init is called once and so is claim. It was one contract per order